File: src/GreenTeaScriptTest.java
//ifdef  JAVA
import java.io.InputStream;
//endif VAJA

class GtScriptRunner {
	public static String LoadFile(String Path) {
		if(LangDeps.HasFile(Path)) {
			return LangDeps.LoadFile(Path);
		}
		return null;
	}
	public static String ExecuteScript(String Path, String Target) {
		/*local*/String[] cmd = {"java", "-jar", "GreenTeaScript.jar", "--" + Target, Path};
		/*local*/String Result = "";
		//FIXME
//ifdef JAVA
		try {
			/*local*/Process proc = new ProcessBuilder(cmd).start();
			proc.waitFor();
			if(proc.exitValue() != 0) {
				return null;
			}
			/*local*/InputStream stdout = proc.getInputStream();
			/*local*/byte[] buffer = new byte[512];
			/*local*/int read = 0;
			while(read > -1) {
				read = stdout.read(buffer, 0, buffer.length);
				if(read > -1) {
					Result += new String(buffer, 0, read);
				}
			}
		} catch (Exception e) {
			e.printStackTrace();
			throw new RuntimeException();
		}
//endif VAJA
		return Result;
	}

	public static void Test(String Target, String ScriptPath, String ResultPath) {
		//LangDeps.println("Testing " + ScriptPath + " (Target:" + Target + ") ... ");
		/*local*/String Expected = GtScriptRunner.LoadFile(ResultPath);
		/*local*/String Actual   = GtScriptRunner.ExecuteScript(ScriptPath, Target);
		LangDeps.Assert(Expected.equals(Actual));
		//LangDeps.println("Testing " + ScriptPath + " (Target:" + Target + ") ... OK");
	}
}

public class GreenTeaScriptTest {
	public static void TestToken(GtContext Context, String Source, String[] TokenTestList) {
		/*local*/GtNameSpace NameSpace = Context.DefaultNameSpace;
		/*local*/GtTokenContext TokenContext = new GtTokenContext(NameSpace, Source, 1);
		/*local*/int i = 0;
		while(i < TokenTestList.length) {
			/*local*/String TokenText = TokenTestList[i];
			LangDeps.Assert(TokenContext.MatchToken(TokenText));
			i = i + 1;
		}
	}

	public static GtContext CreateContext() {
		/*local*/String CodeGeneratorName = "Java";
		/*local*/GtGenerator Generator = LangDeps.CodeGenerator(CodeGeneratorName);
		return new GtContext(new DScriptGrammar(), Generator);
	}

	public static void TokenizeOperator0() {
		GtContext Context = GreenTeaScriptTest.CreateContext();
		/*local*/String[] TokenTestList0 = {"1", "||", "2"};
		GreenTeaScriptTest.TestToken(Context, "1 || 2", TokenTestList0);

		/*local*/String[] TokenTestList1 = {"1", "==", "2"};
		GreenTeaScriptTest.TestToken(Context, "1 == 2", TokenTestList1);

		/*local*/String[] TokenTestList2 = {"1", "!=", "2"};
		GreenTeaScriptTest.TestToken(Context, "1 != 2", TokenTestList2);

		/*local*/String[] TokenTestList3 = {"1", "*", "=", "2"};
		GreenTeaScriptTest.TestToken(Context, "1 *= 2", TokenTestList3);

		/*local*/String[] TokenTestList4 = {"1", "=", "2"};
		GreenTeaScriptTest.TestToken(Context, "1 = 2", TokenTestList4);
	}
	
	public static void TokenizeStatement() {
		GtContext Context = GreenTeaScriptTest.CreateContext();
		/*local*/String[] TokenTestList0 = {"int", "+", "(", "int", "x", ")", ";"};
		GreenTeaScriptTest.TestToken(Context, "int + (int x);", TokenTestList0);
	}

	public static void main(String[] args) {
		if(args.length != 3) {
			GreenTeaScriptTest.TokenizeOperator0();
			GreenTeaScriptTest.TokenizeStatement();
		} else {
			GtScriptRunner.Test(args[0], args[1], args[2]);
		}
	}
}

File: src/TypeScript/GreenTeaScriptTest.js
var GtScriptRunner = (function () {
    function GtScriptRunner() {
    }
    GtScriptRunner.LoadFile = function (Path) {
        if (LangDeps.HasFile(Path)) {
            return LangDeps.LoadFile(Path);
        }
        return null;
    };
    GtScriptRunner.ExecuteScript = function (Path, Target) {
        var cmd = ["java", "-jar", "GreenTea.jar", "--" + Target, Path];
        var Result = "";

        return Result;
    };

    GtScriptRunner.Test = function (Target, ScriptPath, ResultPath) {
        console.log("Testing " + ScriptPath + " (Target:" + Target + ") ... ");
        var Expected = GtScriptRunner.LoadFile(ResultPath);
        var Actual = GtScriptRunner.ExecuteScript(ScriptPath, Target);
        LangDeps.Assert(Expected.equals(Actual));
        console.log("Testing " + ScriptPath + " (Target:" + Target + ") ... OK");
    };
    return GtScriptRunner;
})();

var GreenTeaScriptTest = (function () {
    function GreenTeaScriptTest() {
    }
    GreenTeaScriptTest.TestToken = function (Context, Source, TokenTestList) {
        var NameSpace = Context.DefaultNameSpace;
        var TokenContext = new GtTokenContext(NameSpace, Source, 1);
        var i = 0;
        while (i < TokenTestList.length) {
            var TokenText = TokenTestList[i];
            LangDeps.Assert(TokenContext.MatchToken(TokenText));
            i = i + 1;
        }
    };

    GreenTeaScriptTest.CreateContext = function () {
        var CodeGeneratorName = "Java";
        var Generator = LangDeps.CodeGenerator(CodeGeneratorName);
        return new GtContext(new DScriptGrammar(), Generator);
    };

    GreenTeaScriptTest.TokenizeOperator0 = function () {
        var Context = GreenTeaScriptTest.CreateContext();
        var TokenTestList0 = ["1", "||", "2"];
        GreenTeaScriptTest.TestToken(Context, "1 || 2", TokenTestList0);

        var TokenTestList1 = ["1", "==", "2"];
        GreenTeaScriptTest.TestToken(Context, "1 == 2", TokenTestList1);

        var TokenTestList2 = ["1", "!=", "2"];
        GreenTeaScriptTest.TestToken(Context, "1 != 2", TokenTestList2);

        var TokenTestList3 = ["1", "*", "=", "2"];
        GreenTeaScriptTest.TestToken(Context, "1 *= 2", TokenTestList3);

        var TokenTestList4 = ["1", "=", "2"];
        GreenTeaScriptTest.TestToken(Context, "1 = 2", TokenTestList4);
    };

    GreenTeaScriptTest.TokenizeStatement = function () {
        var Context = GreenTeaScriptTest.CreateContext();
        var TokenTestList0 = ["number", "+", "(", "number", "x", ")", ";"];
        GreenTeaScriptTest.TestToken(Context, "number + (x: number);", TokenTestList0);
    };

    GreenTeaScriptTest.main = function (args) {
        if (args.length != 3) {
            GreenTeaScriptTest.TokenizeOperator0();
            GreenTeaScriptTest.TokenizeStatement();
        } else {
            GtScriptRunner.Test(args[0], args[1], args[2]);
        }
    };
    return GreenTeaScriptTest;
})();
